ecosystem-modules/chainsafe/src/lib.rs

```diff
@@ -0,0 +1,214 @@
+			T::RegistorOrigin::ensure_origin(origin)?;
+			if let Some(currency_id) = CurrencyIds::<T>::take(resource_id) {
+				ResourceIds::<T>::remove(currency_id);
+				Self::deposit_event(Event::UnregisterResourceId(resource_id, currency_id));
+			}
+			Ok(().into())
+		}
+
+		#[pallet::weight(1_000_000)]
+		#[transactional]
+		pub fn transfer_to_bridge(
+			origin: OriginFor<T>,
+			currency_id: CurrencyId,
+			dest_chain_id: chainbridge::ChainId,
+			recipient: Vec<u8>,
+			amount: Balance,
+		) -> DispatchResultWithPostInfo {
+			let who = ensure_signed(origin)?;
+			Self::do_transfer_to_bridge(&who, currency_id, dest_chain_id, recipient, amount)?;
+			Ok(().into())
+		}
+
+		#[pallet::weight(1_000_000)]
+		#[transactional]
+		pub fn transfer_native_to_bridge(
+			origin: OriginFor<T>,
+			dest_chain_id: chainbridge::ChainId,
+			recipient: Vec<u8>,
+			amount: Balance,
+		) -> DispatchResultWithPostInfo {
+			let who = ensure_signed(origin)?;
+			Self::do_transfer_to_bridge(&who, T::NativeCurrencyId::get(), dest_chain_id, recipient, amount)?;
+			Ok(().into())
+		}
+
+		#[pallet::weight(1_000_000)]
+		#[transactional]
+		pub fn transfer_from_bridge(
+			origin: OriginFor<T>,
+			to: T::AccountId,
+			amount: Balance,
+			resource_id: ResourceId,
+		) -> DispatchResultWithPostInfo {
+			let bridge_account_id = T::BridgeOrigin::ensure_origin(origin)?;
+			let currency_id = Self::currency_ids(resource_id).ok_or(Error::<T>::ResourceIdNotRegistered)?;
+
+			if Self::is_origin_chain_resource(resource_id) {
+				// transfer locked tokens from bridge account to receiver
+				T::Currency::transfer(currency_id, &bridge_account_id, &to, amount)?;
+			} else {
+				// issue tokens to receiver
+				T::Currency::deposit(currency_id, &to, amount)?;
+			}
+
+			Ok(().into())
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	fn do_transfer_to_bridge(
+		from: &T::AccountId,
+		currency_id: CurrencyId,
+		dest_chain_id: chainbridge::ChainId,
+		recipient: Vec<u8>,
+		amount: Balance,
+	) -> DispatchResult {
+		ensure!(
+			chainbridge::Module::<T>::chain_whitelisted(dest_chain_id),
+			Error::<T>::InvalidDestChainId
+		);
+
+		let bridge_account_id = chainbridge::Module::<T>::account_id();
+		let resource_id = Self::resource_ids(currency_id).ok_or(Error::<T>::ResourceIdNotRegistered)?;
+
+		if Self::is_origin_chain_resource(resource_id) {
+			// transfer tokens to bridge account to lock
+			T::Currency::transfer(currency_id, &from, &bridge_account_id, amount)?;
+		} else {
+			// burn tokens
+			T::Currency::withdraw(currency_id, &from, amount)?;
+		}
+
+		chainbridge::Module::<T>::transfer_fungible(
+			dest_chain_id,
+			resource_id,
+			recipient,
+			sp_core::U256::from(amount.saturated_into::<u128>()),
+		)
+	}
+
+	fn is_origin_chain_resource(resource_id: ResourceId) -> bool {
+		let origin_chain_id = <T as chainbridge::Config>::ChainId::get();
+		resource_id[31] == origin_chain_id
```

Comment:
is possible for this to panic due to access invalid index?

Reply:
`ResourceId` is alias of `[u8; 32]`